### PR TITLE
refactor(ui): refactor ui form

### PR DESCRIPTION
**Pull request by simeng-li**
## Summary

Create useForm hook

Refactor all sign-in / register forms using global useForm hooks. Extract common logic and handle the input field update and validation logic universally. 

## Linear Issue Reference

LOG-2092

## Testing

ut case passed
test locally
@logto-io/eng

### CODE REVIEW

File: packages/ui/src/utils/field-validations.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+const usernameRegx = /^[A-Z_a-z-][\w-]*$/;
+const emailRegEx = /^\S+@\S+\.\S+$/;
+
+export const usernameValidation = (username: string): ErrorType | undefined => {
+  if (!username) {
+    return { code: 'required', data: { field: i18next.t('input.username') } };
+  }
+
+  if (/\d/.test(username.slice(0, 1))) {
```
Comment (IceHe) [inline packages/ui/src/utils/field-validations.ts:13]: better `/^\d/.test(username)`

(`^` asserts position at start of a line)

Comment (simeng-li) [reply packages/ui/src/utils/field-validations.ts:13]: should be fine with a single digit?

Comment (xiaoyijun) [reply packages/ui/src/utils/field-validations.ts:13]: need `^`
image

**Review by xiaoyijun — CHANGES_REQUESTED** [review]

File: packages/ui/src/containers/CreateAccount/index.tsx [hunk 2]
```diff
@@ -26,119 +30,46 @@ type FieldState = {
...
-
-      return;
-    }
-
-    const confirmPasswordError = validations.confirmPassword?.(fieldState);
-
-    if (confirmPasswordError) {
-      setFieldErrors((previous) => ({ ...previous, confirmPassword: confirmPasswordError }));
-
+    if (!formValidation()) {
```
Comment (gao-sun) [inline packages/ui/src/containers/CreateAccount/index.tsx:56]:
```suggestion
    if (!validateForm()) {
```
use verb for function names?

**Review by xiaoyijun — APPROVED** [review]
LGTM

File: packages/ui/src/utils/field-validations.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import i18next from 'i18next';
+
+import { ErrorType } from '@/components/ErrorMessage';
+
+const usernameRegx = /^[A-Z_a-z-][\w-]*$/;
+const emailRegEx = /^\S+@\S+\.\S+$/;
```
Comment (gao-sun) [inline packages/ui/src/utils/field-validations.ts:6]: align naming to `Regex` or `RegEx`

File: packages/ui/src/hooks/use-form.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  // Revalidate on Input change
+  useEffect(() => {
+    setFieldErrors((previous) => {
+      const errors = entries(fieldValue).map<[keyof T, ErrorType | undefined]>(([key, value]) => [
+        key,
+        // Only validate field with existing errors
+        previous[key] && fieldValidationsRef.current[key]?.(value),
+      ]);
+
+      return Object.fromEntries(errors) as ErrorState;
```
Comment (gao-sun) [inline packages/ui/src/hooks/use-form.ts:57]: just curious why do we still need as here?

File: packages/ui/src/utils/field-validations.ts (hunk 3 above)
Comment (simeng-li) [reply packages/ui/src/utils/field-validations.ts:6]: more like a typo to me 🙆‍♂️

Comment (simeng-li) [reply packages/ui/src/utils/field-validations.ts]: updated

File: packages/ui/src/hooks/use-form.ts (hunk 4 above)
Comment (simeng-li) [reply packages/ui/src/hooks/use-form.ts:57]: aligned. Object.fromEntries will remove the key types

File: packages/ui/src/utils/field-validations.ts (hunk 1 above)
Comment (simeng-li) [reply packages/ui/src/utils/field-validations.ts]: updated

**Pull request merged**